### DISCUSSION TITLE
Single Table Inheritance - WHERE generation depends on SELECT contents

### DISCUSSION THREAD

**Discussion by John-Sias** [sqlalchemy/sqlalchemy#8720] [Usage Questions] [answered]
I've used the Single Table Inheritance model to build an ORM for an existing SQL database with "polymorphic" tables relating to tables of various other object_types with two-field keys consisting of <object_type, object_id>. 

I am having the issue that when I build a query based on a select_from of one of the "child" tables (which should involve a query of the parent under a WHERE for a given polymorphic_identity), SQLAlchemy only generates the WHERE when the child table is included in the select (but not when it is included only in the select_from).

I can reproduce the issue in the standard Single Table Inheritance example as follows:
```
from sqlalchemy import Column, Integer, String, ForeignKey, select
from sqlalchemy.orm import declarative_base, relationship

Base = declarative_base()

class Company(Base):
    __tablename__ = "company"
    id = Column(Integer, primary_key=True)
    name = Column(String(50))
    employees = relationship("Employee", back_populates="company")

class Employee(Base):
    __tablename__ = "employee"
    id = Column(Integer, primary_key=True)
    name = Column(String(50))
    type = Column(String(50))
    company_id = Column(ForeignKey("company.id"))
    company = relationship("Company", back_populates="employees")

    __mapper_args__ = {
        "polymorphic_identity": "employee",
        "polymorphic_on": type,
    }

class Manager(Employee):
    manager_data = Column(String(50))

    __mapper_args__ = {
        "polymorphic_identity": "manager",
    }

class Engineer(Employee):
    engineer_info = Column(String(50))

    __mapper_args__ = {
        "polymorphic_identity": "engineer",
    }
```

The following selects, as we'd expect, only from Managers (WHERE employee.type IN ...)
```
def prst(st):
    print(st.compile(compile_kwargs={'literal_binds':True}))

stmt1 = select(Company.id, Manager.name).select_from(Manager).join(Manager.company)
prst(stmt1)
```
**SELECT company.id, employee.name 
FROM employee JOIN company ON company.id = employee.company_id 
WHERE employee.type IN ('manager')**

But when Manager is not in the select, the WHERE is lost and we get companies related to any Employee (not just Managers)

```
stmt2 = select(Company.id).select_from(Manager).join(Manager.company)
prst(stmt2)
```

**SELECT company.id 
FROM employee JOIN company ON company.id = employee.company_id**

I would expect to see the "WHERE employee.type IN ('manager')" here, as well. Am I missing something or is this a problem in single table polymorphic select compilation?

Thanks in advance for any ideas!

John

**Comment by zzzeek** [comment, marked as answer]
straight up bug, 1.4 and 2.0.  added in #8721 will be fixed in 1.4.43, thanks for reporting

**Reply by John-Sias** [reply]
Fantastic! Thanks @zzzeek!
